warehouse/query-core/src/main/java/datawave/query/jexl/visitors/MergeAdjacentJunctionsVisitor.java

```diff
@@ -0,0 +1,124 @@
+
+/**
+ * A visitor that merges adjacent {@link ASTAndNode} and {@link ASTOrNode}
+ * <p>
+ * It is easier to run multiple visitors in sequence than it is to merge all functionality into a single visitor.
+ */
+public class MergeAdjacentJunctionsVisitor extends BaseVisitor {
+
+    public static JexlNode merge(JexlNode node) {
+        MergeAdjacentJunctionsVisitor visitor = new MergeAdjacentJunctionsVisitor();
+        node.jjtAccept(visitor, null);
+        return node;
+    }
+
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        node.childrenAccept(this, data);
+        mergeSimilarChildren(node);
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        node.childrenAccept(this, data);
+        mergeSimilarChildren(node);
+        return data;
+    }
+
+    /**
+     * Merges any children of the provided JexlNode that match the node id
+     *
+     * @param node
+     *            the JexlNode
+     */
+    protected void mergeSimilarChildren(JexlNode node) {
+
+        if (QueryPropertyMarker.findInstance(node).isAnyType()) {
+            // do not operate on any marker type
+            return;
+        }
+
+        boolean adjacent = false;
+        int target = JexlNodes.id(node);
+
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode child = node.jjtGetChild(i);
+            if (JexlNodes.id(child) == target && !QueryPropertyMarker.findInstance(child).isAnyType()) {
+                adjacent = true;
+                break;
+            }
+        }
+
+        if (adjacent) {
+            List<JexlNode> children = new ArrayList<>();
+            for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+                JexlNode child = node.jjtGetChild(i);
+                if (JexlNodes.id(child) == target && !QueryPropertyMarker.findInstance(child).isAnyType()) {
+                    children.addAll(List.of(JexlNodes.getChildren(child)));
+                } else {
+                    children.add(child);
+                }
+            }
+            JexlNodes.setChildren(node, children.toArray(new JexlNode[0]));
+        } else if (node.jjtGetNumChildren() == 1 && node.jjtGetParent() != null) {
+            // junction with single child should be eliminated
+            JexlNodes.swap(node.jjtGetParent(), node, node.jjtGetChild(0));
+        }
+    }
+
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        super.visit(node, data);
+
+        if (node.jjtGetParent() != null && node.jjtGetNumChildren() == 1) {
+
+            JexlNode parent = node.jjtGetParent();
+            JexlNode child = node.jjtGetChild(0);
+
+            if (areParentAndChildSameJunctions(parent, child)) {
+                List<JexlNode> children = new ArrayList<>();
+                for (int i = 0; i < parent.jjtGetNumChildren(); i++) {
+                    if (parent.jjtGetChild(i) == node) {
```

Comment:
Noticed multiple calls to `parent.jjtGetChild(i)`. Would probably be easier to read if you had a scoped variable such as `currChild` that is instantiated once and then used.

Reply:
Agree